x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Bucket.java

```diff
@@ -89,41 +94,49 @@ public class Bucket extends EsqlScalarFunction implements Validatable {
     public Bucket(
         Source source,
         @Param(name = "field", type = { "integer", "long", "double", "date" }) Expression field,
-        @Param(name = "buckets", type = { "integer" }) Expression buckets,
-        @Param(name = "from", type = { "integer", "long", "double", "date", "keyword", "text" }) Expression from,
-        @Param(name = "to", type = { "integer", "long", "double", "date", "keyword", "text" }) Expression to
+        @Param(name = "bucketsOrSpan", type = { "integer", "double", "date_period", "time_duration" }) Expression bucketsOrSpan,
+        @Param(name = "from", type = { "integer", "long", "double", "date", "keyword", "text" }, optional = true) Expression from,
+        @Param(name = "to", type = { "integer", "long", "double", "date", "keyword", "text" }, optional = true) Expression to
     ) {
-        super(source, List.of(field, buckets, from, to));
+        super(source, from != null && to != null ? List.of(field, bucketsOrSpan, from, to) : List.of(field, bucketsOrSpan));
         this.field = field;
-        this.buckets = buckets;
+        this.bucketsOrSpan = bucketsOrSpan;
         this.from = from;
         this.to = to;
     }
 
     @Override
     public boolean foldable() {
-        return field.foldable() && buckets.foldable() && from.foldable() && to.foldable();
+        return field.foldable() && bucketsOrSpan.foldable() && (from == null || from.foldable()) && (to == null || to.foldable());
     }
 
     @Override
     public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
-        int b = ((Number) buckets.fold()).intValue();
-
         if (field.dataType() == DataTypes.DATETIME) {
-            long f = foldToLong(from);
-            long t = foldToLong(to);
-            return DateTrunc.evaluator(
-                source(),
-                toEvaluator.apply(field),
-                new DateRoundingPicker(b, f, t).pickRounding().prepareForUnknown()
-            );
+            Rounding.Prepared preparedRounding;
+            if (bucketsOrSpan.dataType().isInteger()) {
+                int b = ((Number) bucketsOrSpan.fold()).intValue();
+                long f = foldToLong(from);
+                long t = foldToLong(to);
+                preparedRounding = new DateRoundingPicker(b, f, t).pickRounding().prepareForUnknown();
+            } else {
+                preparedRounding = DateTrunc.createRounding(bucketsOrSpan.fold(), DEFAULT_TZ);
+            }
+            return DateTrunc.evaluator(source(), toEvaluator.apply(field), preparedRounding);
```

Comment:
> make it replace BUCKET(datefield, span) simply by DATE_TRUNC(datefield, span)

It might work for dates, but what about numeric bucketing?

Reply:
We could either keep it the same - or also replace it with the expression currently given in `toEvaluator`, i.e.
`new Mul(source(), floor, rounding)` (where the `rounding` and `floor` can be folded at constant folding time, rather than manually folding in `toEvaluator`).

That _does_ increase the scope a bit for the sake of reducing tech debt, so if at all we should do this in a follow-up.

Reply:
Err, sorry - picking the correct rounding will not work this way. But that's okay, this makes sense to keep as special logic of `BUCKET` - but in the date case, the logic is duplicated in `DATE_TRUNC`, I think.

Reply:
The original `AUTO_BUCKET` needs the logic within the function itself, since both date and numeric spans are computed. The newly added functionality of user-provided spans could be surogate-substituted entirely, but we'd spread the logic. The date case is pretty much a 1-to-1 mapping, the numerical one isn't. Not sure there's an upside, but not absolutely opposing the idea either.